eth2/utils/eth2_wallet/src/wallet.rs

```diff
@@ -0,0 +1,287 @@
+        } else {
+            let salt = rand::thread_rng().gen::<[u8; SALT_SIZE]>();
+            let iv = rand::thread_rng().gen::<[u8; IV_SIZE]>().to_vec().into();
+
+            Ok(Self {
+                seed: seed.to_vec().into(),
+                password,
+                kdf: default_kdf(salt.to_vec()),
+                cipher: Cipher::Aes128Ctr(Aes128Ctr { iv }),
+                uuid: Uuid::new_v4(),
+                nextaccount: 0,
+                name,
+            })
+        }
+    }
+
+    /// Consumes `self`, returning an encrypted `Wallet`.
+    pub fn build(self) -> Result<Wallet, Error> {
+        Wallet::encrypt(
+            self.seed.as_bytes(),
+            self.password,
+            self.kdf,
+            self.cipher,
+            self.uuid,
+            self.name,
+            self.nextaccount,
+        )
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(transparent)]
+pub struct Wallet {
+    json: JsonWallet,
+}
+
+impl Wallet {
+    /// Instantiates `Self`, encrypting the `seed` using `password` (via `kdf` and `cipher`).
+    ///
+    /// The `uuid`, `name` and `nextaccount` are carried through into the created wallet.
+    fn encrypt(
+        seed: &[u8],
+        password: &[u8],
+        kdf: Kdf,
+        cipher: Cipher,
+        uuid: Uuid,
+        name: String,
+        nextaccount: u32,
+    ) -> Result<Self, Error> {
+        let (cipher_text, checksum) = encrypt(&seed, &password, &kdf, &cipher)?;
+
+        Ok(Self {
+            json: JsonWallet {
+                crypto: Crypto {
+                    kdf: KdfModule {
+                        function: kdf.function(),
+                        params: kdf,
+                        message: EmptyString,
+                    },
+                    checksum: ChecksumModule {
+                        function: Sha256Checksum::function(),
+                        params: EmptyMap,
+                        message: checksum.to_vec().into(),
+                    },
+                    cipher: CipherModule {
+                        function: cipher.function(),
+                        params: cipher,
+                        message: cipher_text.into(),
+                    },
+                },
+                uuid,
+                nextaccount,
+                version: Version::one(),
+                type_field: TypeField::Hd,
+                name,
+            },
+        })
+    }
+
+    /// Produces a `Keystore` (encrypted with `keystore_password`) for the validator at
+    /// `self.nextaccount`, incrementing `self.nextaccount` if the keystore was successfully
+    /// generated.
+    ///
+    /// Uses the default encryption settings of `KeystoreBuilder`, not necessarily those that were
+    /// used to encrypt `self`.
+    ///
+    /// ## Errors
+    ///
+    /// - If `wallet_password` is unable to decrypt `self`.
+    /// - If `keystore_password.is_empty()`.
+    /// - If `self.nextaccount == u32::max_value() - 1`.
```

Comment:
`self.nextaccount == u32::max_value()` ?

The `checked_add()` would fail when we add 1 to `u32::max_value`?